File: 04-Data-Structures/code/typescript/src/ringBuffer.ts
import { DSError } from "./errors.js";

/** Fixed-capacity circular queue: O(1) push/pop with no shifting. */
export class RingBuffer<T> {
  private data: Array<T | undefined>;
  private head = 0;
  private count = 0;
  private cap: number;

  constructor(capacity: number) {
    if (capacity <= 0) throw new DSError("invalid", "capacity must be > 0");
    this.cap = capacity;
    this.data = new Array<T | undefined>(capacity);
  }

  size(): number {
    return this.count;
  }

  isFull(): boolean {
    return this.count === this.cap;
  }

  isEmpty(): boolean {
    return this.count === 0;
  }

  push(x: T): void {
    if (this.isFull()) throw new DSError("full", "RingBuffer is full");
    const tail = (this.head + this.count) % this.cap;
    this.data[tail] = x;
    this.count++;
  }

  pop(): T {
    if (this.isEmpty()) throw new DSError("empty", "RingBuffer is empty");
    const value = this.data[this.head] as T;
    this.data[this.head] = undefined;
    this.head = (this.head + 1) % this.cap;
    this.count--;
    return value;
  }

  toList(): T[] {
    const out: T[] = [];
    for (let i = 0; i < this.count; i++) out.push(this.data[(this.head + i) % this.cap] as T);
    return out;
  }

  checkInvariants(): void {
    if (this.count < 0 || this.count > this.cap) throw new DSError("invalid", "count out of range");
    if (this.head < 0 || this.head >= this.cap) throw new DSError("invalid", "head out of range");
  }
}

File: 04-Data-Structures/code/python/seb_ds/ring_buffer.py
"""Fixed-capacity circular queue. Mirrors `typescript/src/ringBuffer.ts`."""

from __future__ import annotations

from typing import Generic, List, TypeVar

from .errors import DSError

T = TypeVar("T")


class RingBuffer(Generic[T]):
    def __init__(self, capacity: int) -> None:
        if capacity <= 0:
            raise DSError("invalid", "capacity must be > 0")
        self._cap = capacity
        self._data: List[T | None] = [None] * capacity
        self._head = 0
        self._count = 0

    def size(self) -> int:
        return self._count

    def is_full(self) -> bool:
        return self._count == self._cap

    def is_empty(self) -> bool:
        return self._count == 0

    def push(self, x: T) -> None:
        if self.is_full():
            raise DSError("full", "RingBuffer is full")
        tail = (self._head + self._count) % self._cap
        self._data[tail] = x
        self._count += 1

    def pop(self) -> T:
        if self.is_empty():
            raise DSError("empty", "RingBuffer is empty")
        value = self._data[self._head]
        self._data[self._head] = None
        self._head = (self._head + 1) % self._cap
        self._count -= 1
        return value  # type: ignore[return-value]

    def to_list(self) -> List[T]:
        return [self._data[(self._head + i) % self._cap] for i in range(self._count)]  # type: ignore[misc]

    def check_invariants(self) -> None:
        if self._count < 0 or self._count > self._cap:
            raise DSError("invalid", "count out of range")
        if self._head < 0 or self._head >= self._cap:
            raise DSError("invalid", "head out of range")
